crates/ruff_cli/src/diagnostics.rs

```diff
@@ -1,35 +1,42 @@
 #![cfg_attr(target_family = "wasm", allow(dead_code))]
 
+use std::collections::HashMap;
 use std::fs::write;
 use std::io;
 use std::io::Write;
 use std::ops::AddAssign;
 use std::path::Path;
 
-use anyhow::Result;
+use anyhow::{anyhow, Result};
 use colored::Colorize;
 use log::{debug, error};
 use rustc_hash::FxHashMap;
 use similar::TextDiff;
 
+use ruff::jupyter::{read_jupyter_notebook, JupyterIndex, JupyterNotebook};
 use ruff::linter::{lint_fix, lint_only, FixTable, LinterResult};
 use ruff::message::Message;
 use ruff::settings::{flags, AllSettings, Settings};
-use ruff::{fix, fs};
+use ruff::{fix, fs, jupyter};
 
 use crate::cache;
+use ruff::jupyter::is_jupyter_notebook;
 
 #[derive(Debug, Default)]
 pub struct Diagnostics {
     pub messages: Vec<Message>,
     pub fixed: FxHashMap<String, FixTable>,
+    /// Jupyter notebook indexing table for each input file that is a jupyter notebook
+    /// so we can rewrite the diagnostics in the end
+    pub jupyter_index: HashMap<String, JupyterIndex>,
```

Comment:
Rather than passing around the index... I'm wondering if we should instead change `Message` to have a more flexible location type, and store the location data on `Message` directly. This isn't a fully-formed idea, but it seems non-ideal that clients of `Diagnostics` need to do this translation, worry about what's a notebook, what isn't, etc. Could `Message` instead have a `Location` that's an enum or similar, and can be (e.g.) a Python file location or a Jupyter notebook location? What would be the tradeoffs?

In other words, this feels like exposing an implementation detail, since clients only care about the location, and not that translation step. What do you think?

Reply:
i'm not completely sure either; my current thought is that most code can just assume to see one consistent python file and not bother with everything else and a special case such as jupyter notebooks gets patched in the appropriate places.

making `location` and `end_location` an enum of `Location(Location)` and `JupyterLocation { cell: u32, row: u32, column: u32}` would definitely also be possible